### ISSUE TITLE
packed_tex Load Failure Using Godot Server Executable 

### ISSUE THREAD

**Issue by kruffin** [Zylann/godot_heightmap_plugin#298]
**Describe the bug**
When using the Godot server executable to run with a HTerrain node and the Classic4 shader it will fail trying to load the packed_tex files.

```
Recreating hterrain_server_terrain-server_1 ... done
Attaching to hterrain_server_terrain-server_1
terrain-server_1  | ERROR: Failed loading resource: res://texture_data/HterrainServer_slot0_albedo_bump.packed_tex. Make sure resources have been imported by opening the project in the editor at least once.
terrain-server_1  |    at: _load (core/io/resource_loader.cpp:270)
terrain-server_1  | ERROR: res://HterrainServer.tscn:6 - Parse Error: [ext_resource] referenced nonexistent resource at: res://texture_data/HterrainServer_slot0_albedo_bump.packed_tex
terrain-server_1  |    at: poll (scene/resources/resource_format_text.cpp:412)
terrain-server_1  | ERROR: Failed to load resource 'res://HterrainServer.tscn'.
terrain-server_1  |    at: load (core/io/resource_loader.cpp:206)
terrain-server_1  | ERROR: Failed loading resource: res://HterrainServer.tscn. Make sure resources have been imported by opening the project in the editor at least once.
terrain-server_1  |    at: _load (core/io/resource_loader.cpp:270)
terrain-server_1  | ERROR: Failed loading scene: res://HterrainServer.tscn
terrain-server_1  |    at: start (main/main.cpp:2011)
terrain-server_1  | WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
terrain-server_1  |      at: cleanup (core/object.cpp:2064)
terrain-server_1  | ERROR: Resources still in use at exit (run with --verbose for details).
terrain-server_1  |    at: clear (core/resource.cpp:417)
terrain-server_1  | ERROR: There are still MemoryPool allocs in use at exit!
terrain-server_1  |    at: cleanup (core/pool_vector.cpp:63)
terrain-server_1  | Godot Engine v3.4.2.stable.mono.official.45eaa2daf - https://godotengine.org
terrain-server_1  |  
terrain-server_1  | Mono: Log file is: '/root/.local/share/godot/app_userdata/hterrain_server/mono/mono_logs/2022-01-18_22.42.46_1.log'
hterrain_server_terrain-server_1 exited with code 0
```

resource_format_text.cpp line 412: https://github.com/godotengine/godot/blob/3.4.2-stable/scene/resources/resource_format_text.cpp#L412
resource_loader.cpp line 206: https://github.com/godotengine/godot/blob/3.4.2-stable/core/io/resource_loader.cpp#L206
resource_loader.cpp line 270: https://github.com/godotengine/godot/blob/3.4.2-stable/core/io/resource_loader.cpp#L270

**To Reproduce**
A minimum project to reproduce is here: https://github.com/kruffin/hterrain_server
Basically, add textures to the terrain node and the packed_tex files will not be loaded by the server binary.

**Expected behavior**
Would be nice to know if this is the expected outcome or how to identify why it's failing to load. I did take a look a the packed texture importer (https://github.com/Zylann/godot_heightmap_plugin/blob/1.5.2/addons/zylann.hterrain/tools/packed_textures/packed_texture_importer.gd) but am wondering if custom importers are broken in Godot Server. Line 270 of resource_loader.cpp makes me think it doesn't have the custom importer, but am not sure.

**Screenshots**
N/A

**Environment**
- OS: Linux
- Graphics card: AMD Radeon RX 6700 XT
- Godot version: 3.4.2
- Plugin version: 1.5.2
- Renderer used: GLES3

**Comment by kruffin** [comment]
Looks like there is a different issue with the reproduction project after cloning new it is expecting some information from the .import folder. So has a dependency missing error even though all the files are there. I had the same error with the demo terrain project. Might have to open a separate issue for this.
![failed_to_load_dependency_2022-01-18 18-29-36](https://user-images.githubusercontent.com/917903/150035717-a1d099c5-d5c0-452e-8116-bd26e2741e46.png)
![dependency_errors_large_2022-01-18 18-32-21](https://user-images.githubusercontent.com/917903/150035750-d8812281-9b80-4693-9f52-398017e0de7f.png)

**Comment by Zylann** [comment]
Having to reproduce this on Linux with a server version including Mono is going to be hard for me, I'll have to try without Mono and on Windows probably, but can't tell when I'll have time to do such investigation.
Also everytime I see a custom loader involved I am suspicious of the feature being randomly broken because it depends on `class_name` which is itself buggy in Godot 3, especially on first project startup.
On top of that, the packed texture importer suffers from [API shortcomings](https://github.com/godotengine/godot-proposals/issues/1943) so any change the engine has made since I wrote them could have caused breakage. It could also be that server builds do something special to textures that I'm not aware of.

**Comment by Zylann** [comment]
> Do I need to commit files in the .import folder to prevent that?

You should not need to commit anything from inside the `.import` folder.

Perhaps we can simplify the problem further by eliminating all terrain stuff. Just have a new project, with a minimal plugin solely defining the packed texture importer. Then have a packed texture inside the project, make a plane mesh use it etc. Then we can see more easily what happens

**Comment by kruffin** [comment]
> Have a new project, with a minimal plugin solely registering the packed texture importer

I'll see if I can put something together tonight.

> You should not need to commit anything from inside the .import folder.

Try opening that sample project (https://github.com/kruffin/hterrain_server/tree/master/godot_files) and note it complains about missing files that exist. I get a similar error with demo project (https://github.com/Zylann/godot_hterrain_demo).

**Comment by Zylann** [comment]
Last time I saw something like [this](https://github.com/Zylann/godot_heightmap_plugin/issues/298#issuecomment-1015930911), it was someone with a failing hard drive, or under Windows a case-insensitive issue. Once again Godot is misreporting errors like a pro, totally saying what's actually going on... will have a look when I have time

**Comment by kruffin** [comment]
Put together that minimal addon and project here: https://github.com/kruffin/packed_tex_test

It has the same error with the server binary, so has to be something with how it treats importers...

**Comment by kruffin** [comment]
So, doing --help on the three different Godot binaries resulted in some interesting options for the video driver.

```
$./packed_tex_test.x86_64 --help
...
--video-driver <driver>          Video driver ('GLES3', 'GLES2').

$./Godot_v3.4.2-stable_linux_headless.64 --help
...
--video-driver <driver>          Video driver ('Dummy').

$./Godot_v3.4.2-stable_linux_headless.64 --help
...
--video-driver <driver>          Video driver ('Dummy').
```

My guess at this point is a video driver is needed to create a stream texture so it complains about the .packed_tex resource when it can't create one. Making this up though and have no proof. It really shouldn't be trying to load the packed texture (none of the files it references are included in the pck).

```
{
	"contains_albedo": true,
	"resolution": 512,
	"src": {
		"a": "res://texture_src/Ground037_1K_Displacement.jpg",
		"rgb": "res://texture_src/Ground037_1K_Color.jpg"
	}
}
```
Slightly more interesting output with the -v parameter when running.

```
$./Godot_v3.4.2-stable_linux_server.64 --main-pack packed_tex_test.pck -v
Godot Engine v3.4.2.stable.official.45eaa2daf - https://godotengine.org
 
CORE API HASH: 0
EDITOR API HASH: 0
Loading resource: res://default_env.tres
Loading resource: res://PackedTexTest.tscn
Loading resource: res://texture_data/HterrainServer_slot0_albedo_bump.packed_tex
ERROR: Failed loading resource: res://texture_data/HterrainServer_slot0_albedo_bump.packed_tex. Make sure resources have been imported by opening the project in the editor at least once.
   at: _load (core/io/resource_loader.cpp:270)
ERROR: res://PackedTexTest.tscn:3 - Parse Error: [ext_resource] referenced nonexistent resource at: res://texture_data/HterrainServer_slot0_albedo_bump.packed_tex
   at: poll (scene/resources/resource_format_text.cpp:412)
ERROR: Failed to load resource 'res://PackedTexTest.tscn'.
   at: load (core/io/resource_loader.cpp:206)
ERROR: Failed loading resource: res://PackedTexTest.tscn. Make sure resources have been imported by opening the project in the editor at least once.
   at: _load (core/io/resource_loader.cpp:270)
ERROR: Failed loading scene: res://PackedTexTest.tscn
   at: start (main/main.cpp:2011)
WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
     at: cleanup (core/object.cpp:2064)
```

```
$ ./packed_tex_test.x86_64 -v
Godot Engine v3.4.2.stable.mono.official.45eaa2daf - https://godotengine.org
XInput: Refreshing devices.
XInput: No touch devices found.
Detecting GPUs, set DRI_PRIME in the environment to override GPU detection logic.
Only one GPU found, using default.
Using GLES3 video driver
OpenGL ES 3.0 Renderer: AMD Radeon RX 6700 XT
OpenGL ES Batching: ON
	OPTIONS
	max_join_item_commands 16
	colored_vertex_format_threshold 0.25
	batch_buffer_size 16384
	light_scissor_area_threshold 1
	item_reordering_lookahead 4
	light_max_join_items 32
	single_rect_fallback False
	debug_flash False
	diagnose_frame False
/lib/x86_64-linux-gnu/libasound.so.2: undefined symbol: snd_dlpath
PulseAudio: context other
PulseAudio: context other
PulseAudio: context other
PulseAudio: context ready
PulseAudio: Detecting channels for device: alsa_output.pci-0000_0d_00.1.hdmi-stereo-extra4
PulseAudio: detected 2 output channels
PulseAudio: audio buffer frames: 512 calculated output latency: 11ms
JoypadLinux: udev enabled and loaded successfully.
 
Mono: Initializing module...
Mono JIT compiler version 6.12.0 ((no/5ce143a1)
ERROR: Mono: Core API hash mismatch.
   at: _init_godot_api_hashes (modules/mono/mono_gd/gd_mono.cpp:490)
Mono: Log file is: '/home/kruffin/.local/share/godot/app_userdata/packed_tex_test/mono/mono_logs/2022-01-21_12.24.49_51615.log'
Mono: Skipping runtime initialization because 'mscorlib.dll' could not be found
CORE API HASH: 12889762389686882091
EDITOR API HASH: 3237768591
Loading resource: res://default_env.tres
Loaded builtin certs
Loading resource: res://PackedTexTest.tscn
Loading resource: res://texture_data/HterrainServer_slot0_albedo_bump.packed_tex
```

Might have to give up on server or headless binaries and just use the --no-window command on the regular one. Probably brings it's own problems with fake frame buffers, but a different rabbit hole.

**Comment by kruffin** [comment]
Maybe something to do with this nonsense: https://github.com/godotengine/godot/blob/3.4.2-stable/drivers/dummy/texture_loader_dummy.cpp#L38

**Comment by Zylann** [comment]
Yeah I was just pointed to the same thing. But I still have no idea what I'm supposed to modify in my plugin for the loading to work, because I did not write any kind of custom loader for this.

**Comment by kruffin** [comment]
Thanks Zylann. Just want to say thank you for putting this height map plugin out there for others to use and play with. I'm going to ask on IRC and see what I can find out. I'll create an issue in the Godot project and report back the URL here.

**Comment by kruffin** [comment]
Thought I should take some time to reply back here with some of the workarounds that can be done to get around this custom resource loading issue on a server executable:

1. Don't load the resource when running from a server binary - seems obvious, but does require some changes to your project depending on how you are instancing the HTerrain node. The gist of that can be seen in a workaround branch of the packed tex test project here: https://github.com/kruffin/packed_tex_test/blob/workaround/PackedTexTest.gd This example isn't exactly the same because you would be setting the `texture_set` attribute on the HTerrain instead, but shows how. Do note the `load()` vs `preload()` because `preload()` will have the same load error.
2. Don't use HTerrain at all on the server and instead use the HTerrainData and HTerrainCollider objects alone. This is the approach I am currently trying and seems to work, however, does require significant changes.
